### ISSUE TITLE
refactor crab resubmit

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6270]
avoid editing dagman log and user new Dagman features developed by Mark C.
Once we have agreed on the semantyc and the new Dagman is available ...

WIll be a cleaner and definitive solution to https://github.com/dmwm/CRABServer/issues/5876

**Comment by belforte** [comment]
Raise priority, time to start looking at this seriously

**Comment by belforte** [comment]
for convenience, paste here  conclusion from #5876 
After discussion with HTCondor developers we came to these conclusion:

1. the problem was not there in 2014 when Brian initially coded this
1. the problem came when condor got smarter about writing logs and stopped locking them by default, opening the way for our log editing procedure (which attempts to use condor file locking API) to overwrite a log file with an "old" version where some events are missing
1. condor can be configured to revert to old behavior by setting `ENABLE_USERLOG_LOCKING=True` in its configuration so that we do not have the problem
1. we made that change in all CRAB schedd's and did not find any sign of increased load or slower operations, so we can run in that way for a while. Ref. https://cms-logbook.cern.ch/elog/Analysis+Operations/3282 
1. editing logs is bad anyhow and it has been agreed to enhance DAGMAN functionality to allow CRAB to do resubmits w/o tampering with files it should not tamper with. Discussion on this has started with HTCondor DAGMAN expert Mark Coatsworth:
 https://docs.google.com/document/d/1vgJApmjkH9brYhQZbdRooGnj2mqSrme7BWVxzReZ0oM/edit

**Comment by belforte** [comment]
will  look at this after transition to py3

**Comment by belforte** [comment]
Update from Mark C.
```

On 22/02/2021 18:40, Mark Coatsworth wrote:
> Hi Stefano, long overdue update on this work (replacing the old CMS
> CRAB log editing mechanism).
> 
> I had a couple false starts but finally implemented the
> DAGMAN_PUT_FAILED_JOBS_ON_HOLD mechanic that we discussed. It's fairly
> simple: when this option is set to True, DAGMan will put failed jobs
> on hold instead of aborting the dag. This gives CRAB the opportunity
> to fix the problem and continue processing the dag (hence, no need to
> edit the log to re-run failed nodes).
> 
> The new feature will ship in the 9.1 release of Condor later this spring.
> 
> Please keep me posted where things are at on your end. I know this
> will involve some changes in the CRAB code, and likely some tweaks on
> the DAGMan side also. I'd be happy to help with this when the time
> comes,
> 
> Mark
```

**Comment by belforte** [comment]
@dciangot since you manifested interest on this, I add you to assignees so you can keep track

**Comment by belforte** [comment]
note also this recent thread in htcondor forum
https://lists.cs.wisc.edu/archive/htcondor-users/2022-February/msg00015.shtml

**Comment by belforte** [comment]
with ref to last line in https://github.com/dmwm/CRABServer/issues/6270#issuecomment-760194316 
2 years later is  better make a copy of the old googleDoc in condor space, just in case:
original: https://docs.google.com/document/d/1vgJApmjkH9brYhQZbdRooGnj2mqSrme7BWVxzReZ0oM/edit
copy in my drive: https://docs.google.com/document/d/1qsil0UGewazg96cA-KP1QVOwkIcjSI6-ceT-qMsOYa0/edit?usp=sharing

**Comment by belforte** [comment]
this should be more straightforward now that we have decided not to allow resubmission of successful jobs https://github.com/dmwm/CRABClient/issues/5285

**Comment by ColeBollig** [comment]
Answering questions about DAGMAN_PUT_FAILED_JOBS_ON_HOLD from email thread here:

1. Do I understand correctly that the (only) way to use it is via a Dagman configuration file?
    Yes. There is no submission equivalent option to enable this behavior and it must be applied via the configuration option either by a DAG specific configuration file or by setting `_CONDOR_DAGMAN_PUT_FAILED_JOBS_ON_HOLD=True` in the environment of the DAGMan subission.
2. What is the proper syntax?
    `DAGMAN_PUT_FAILED_JOBS_ON_HOLD = True`
3. How does this function?
    Looking at the code, this only takes effect on node without a post-script once all retry attempts have been attempted. Once here, a job failure (non-zero exit code or abort event) will cause DAGMan to restart the node with the jobs being placed to the AP in the held state. This specific retrying will keep occurring until the jobs are successful or DAGMan is removed from the AP. Note that the jobs in the AP are not the original jobs that failed but new ones specifically place to the AP in the held state.

**Comment by ColeBollig** [comment]
> overall you make me feel that the design document needs to be revisited

I agree with this statement. Even if the design doesn't change I think the implementation needs to be revised/improved

**Comment by aspiringmind-code** [comment]
@ColeBollig 
We tried using Rescue dags to do resubmission and we are running into some problems and your helpful insight will be highly appreciated.
We have been trying to make use of the configuration variable DAGMAN_RESET_RETRIES_UPON_RESCUE by setting it to True in a config file that we refer to with CONFIG command in the dag file. However, it doesn't seem to make any difference to the Retries. We have a simple RETRY Job <no. of retries> command that we want to reset when the number of retries are done and failure persists leading to writing of rescue file along with some PRE and POST scripts. We also don't find the variable set in the output dagman.out. 

In the backdrop of this problem, I have the following questions:

A.) Is there a conclusive way to make sure our config file is being read and the config variable being set to True besides checking for the variable in the output file? Or an alternative way to set this variable?

B.) The manual mentions the line "If the Rescue DAG file is generated before all retries of a node are completed, then the Rescue DAG file will also contain RETRY entries."

I want to confirm that this "RETRY entries" is  just copying the RETRY lines from the dagman script. For instance, if my DAG file didn't have a RETRY command to begin with, would I still get a RETRY entry in the rescue? 

Secondly, the word "before": Our use case envisions using the latest rescue dag file generated after all the retries are completed for the first time. Once the retries are completed, we do a resubmission that uses the latest rescue file which we want to reset the retries for, so effectively run the number of retries again.  Is that not possible? What we find currently is that rescue files have only the RETRY line (same as that we pass in the actual DAG) It seems that it is copying the RETRY entry from original DAG and the reset never happens.

Also, it's my understanding that the rescue file should copy the original DAG PRE Job and POST lines verbatim in the rescue file from the dag file. However, we find that the rescue file has only the RETRY line after the header comments about failed DAG execution. I suspect since the Rescue file has no other line it is not working as expected.  

Thanks!
Cheers,
Vijay Chakravarty
